File: src/main/java/lmller/github/io/gofkotlin/observer/Observer.java
package lmller.github.io.gofkotlin.observer;

import java.util.ArrayList;
import java.util.List;

public class Observer {
    public interface PropertyChangedListener {
        void onChange(int newValue);
    }

    public static class Property {
        private int value;

        private List<PropertyChangedListener> listeners = new ArrayList<>();

        public void registerListener(PropertyChangedListener listener) {
            listeners.add(listener);
        }

        public void setValue(int value) {
            this.value = value;
            notifyChange();
        }

        private void notifyChange() {
            for (PropertyChangedListener listener : listeners) {
                listener.onChange(value);
            }
        }
    }

    public static void main(String[] args) {
        Property property =  new Property();
        property.registerListener(new PropertyChangedListener() {
            @Override
            public void onChange(int newValue) {
                System.out.println("The new value is " + newValue);
            }
        });

        property.setValue(89);

    }
}

File: src/main/java/lmller/github/io/gofkotlin/observer/Observer.kt
package lmller.github.io.gofkotlin.observer

import kotlin.properties.Delegates

typealias PropertyChangeListener = (Int) -> Unit

class Property {

    private val listeners = arrayListOf<PropertyChangeListener>()

    var value by Delegates.observable(0) { _, old, new ->
        println("The value was: $old")
        listeners.forEach { listener -> listener(new) }
    }

    fun registerListener(listener: PropertyChangeListener) {
        listeners.add(listener)
    }
}

fun main(args: Array<String>) {
    val property = Property()
    property.registerListener { new -> println("The new value is $new") }

    property.value = 1989
    property.value = 2017
}